_posts/2020-01-09-rxjs-wstep.md

```diff
@@ -0,0 +1,180 @@
+Observable jest obiektem reprezentującym strumień danych. Implementuje wzorzec projektowy _Obserwator_, który zakłada istnienie bytu przechowującego listę obiektów - obserwatorów, nasłuchujących na jakiekolwiek zmiany stanu danego bytu, powiadamiającego każdego z nich o zmianie poprzez wywołanie funkcji przez nich przekazanych (callback).
+
+Najprostszego Observable możemy stworzyć za pomocą funkcji statycznej _of_.
+```typescript
+// obiekt Observable, emitujący wartości liczbowe od 1 do 5
+const numbers$: Observable<number> = of(1,2,3,4,5);
+
+```
+Obiekt numbers$ jest definicją strumienia danych typu _number_. Jest to tylko i wyłącznie szablon strumienia. W tym przypadku stworzyliśmy strumień zimny. Znamy zbiór danych (1,2,3,4,5), jednak dane zaczną być emitowane dopiero w momencie rozpoczęcia nasłuchiwania na dany strumień przez obserwatora. Pod strumień 'podłączamy' się za pomocą funkcji subscribe().
+### subscribe() i unsubscribe()
+Funkcja subscribe jako parametr oczekuje obiektu, który definiuje trzy funkcje: _next_, _error_ oraz _complete_.
+```typescript
+const subscription = numbers$.subscribe({
+    next(value) {},
+    error(err) {},
+    complete() {}
+});
+```
+Każda z tych funkcji jest _callbackiem_, który jest wywoływany w poszczególnych momentach przepływu danych przez strumień. Funkcja _next(value)_ wywoływana jest za każdym razem, gdy strumień emituje pojedyńczą wartość - czyli w naszym przypadku funkcja _next()_ zostanie wywołana 5 razy, raz dla każdej z cyfr z zakresu 1-5. Callback _error(err)_ zostanie wywołany w momencie, gdy strumień zostanie nienaturalnie zamknięty lub przerwany. _Complete_ jest ostatnim callbackiem, który wywołoywany jest po zamknięciu strumienia.
+
+Wywołanie funkcji _subscribe()_ dopisuje nas do listy obserwatorów danego strumienia.
+
+W momencie podpięcia się jako obserwator do danego Observable zostaje przekazany nam obiekt typu Subscription, na którym możemy wywołać metodę _unsubscribe()_, która usunie nas z listy obserwatorów. 
+
+Odsubskrybowanie się jest bardzo ważne w przypadku strumieni gorących, które w większości przypadków są nieskończone - emitują wartość przez potencjalnie nieskończoną ilość czasu. Jeżeli zapomnimy o usunięciu nas z listy obserwatorów danego strumienia, referencja do stworzonego przez nas obserwatora będzie istnieć przez cały cykl życia aplikacji - tworząc nieskończoną dziurę w pamięci, która w sytuacji ekstremalnej może doprowadzić do zabicia karty, w której działa nasza aplikacja.
+## Subject - tworzenie własnych strumieni
+RxJS oferuje nam również możliwość tworzenia własnych strumieni. Możemy to zrobić za pomocą obiektów typu Subject. Subject, tłumacząc na język polski, oznacza dosłownie: _temat_. Strumień taki jest więc tematem, potencjalnie nieskończonym, który emituje nowe wartości w kluczowych dla nas miejscach aplikacji. 
```

Comment:
To można łatwo skrócić ==> Możemy to zrobić za pomocą obiektów typu Subject (tłumacząc dosłownie na język polski: _temat_).

Strumień taki jest więc tematem ==> **Taki strumień** jest więc tematem